# WFM Demand Forecasting Model

**Interval-level intraday call volume forecasting, shrinkage modeling, Erlang C service-level estimation, and FTE gap analysis for contact-center staffing optimization.**

---

This notebook demonstrates an end-to-end Workforce Management (WFM) forecasting and capacity-planning methodology applied to a real call-center arrival dataset. The workflow mirrors the daily and intraday planning cycle a WFM analyst runs in a production contact center:

1. **Load & shape** real interval-level call arrival data
2. **Forecast** call volume at the interval level (intraday seasonality + day-of-week patterns)
3. **Apply shrinkage** to translate productive workload into a real staffing requirement
4. **Size the workforce** with Erlang C to hit a target service level
5. **Run gap analysis** comparing required vs. scheduled FTEs and flag over/under-staffed intervals

**Author's note on provenance.** The business logic, WFM domain framing, and operational assumptions in this notebook (shrinkage composition, service-level targets, occupancy ceilings, FTE conversion, staffing-gap interpretation) reflect my own contact-center workforce-management experience. AI tooling assisted with Python syntax and plotting code. The arrival data is a real, published academic dataset; the WFM overlay (shrinkage rates, AHT assumptions, schedule) is synthetic and clearly labeled as such wherever it appears.


## 1. The Dataset

**Source:** "Anonymous Bank" call-center data, Service Enterprise Engineering (SEE) Lab, Technion — Israel Institute of Technology. Originally documented by Guedj & Mandelbaum (2000); widely used in the queueing-theory and call-center literature (Brown et al., 2005).

Accessed via the `Rfssa` R package data mirror, which publishes the arrivals pre-aggregated into **6-minute intervals**:

> `https://github.com/haghbinh/dataset/raw/main/Rfssa_dataset/Callcenter.rds`

The data covers **January 1 – December 31, 1999** for a small retail bank call center. Each row is one 6-minute interval on one day:

| column | meaning |
|---|---|
| `calls` | number of calls arriving in that 6-minute interval |
| `u` | interval index within the day, 1–240 (240 × 6 min = 24 h) |
| `Date` | date key |
| `Day` | day of week |
| `Month` | month |

**Why this dataset for a WFM demonstration.** It provides genuine interval-level arrival counts — the raw material of every contact-center forecast — with the seasonality structure (intraday peaks, day-of-week effects) that staffing models must capture. Because it is a *Sunday–Thursday* business week (an Israeli institution), Friday/Saturday are the low-volume weekend; the methodology is identical for a Mon–Fri operation.

**Acknowledgement.** Per the data owners' request: data source is the SEE Lab at the Technion (Prof. Avi Mandelbaum). The data is free for academic/non-commercial use.


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick
import seaborn as sns
from math import factorial

sns.set_style("whitegrid")
plt.rcParams["figure.figsize"] = (12, 5)
plt.rcParams["axes.titlesize"] = 13
plt.rcParams["axes.titleweight"] = "bold"
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")
RNG = np.random.default_rng(42)  # reproducible synthetic WFM overlay


### 1.1 Load the data

The file is an R `.rds`. We read it with `pyreadr` (no R install needed). If you are running this notebook fresh, the cell will download the file to a local `data/` folder on first run.


In [2]:
import os
try:
    import pyreadr
except ImportError:
    import sys, subprocess
    subprocess.run([sys.executable, "-m", "pip", "install", "pyreadr", "-q"])
    import pyreadr

import urllib.request

DATA_URL = "https://github.com/haghbinh/dataset/raw/main/Rfssa_dataset/Callcenter.rds"
os.makedirs("data", exist_ok=True)
local_path = "Callcenter.rds" if os.path.exists("Callcenter.rds") else "data/Callcenter.rds"   # bundled copy first
if not os.path.exists(local_path):
    print("Downloading Callcenter.rds ...")
    urllib.request.urlretrieve(DATA_URL, local_path)

raw = pyreadr.read_r(local_path)[None]
print("Rows:", len(raw), "| Columns:", list(raw.columns))
raw.head()


Rows: 87600 | Columns: ['calls', 'u', 'Date', 'Day', 'Month']


,calls,u,Date,Day,Month
0,2.00,1,1970-01-01,Friday,Jan
1,0.00,2,1970-01-01,Friday,Jan
2,0.00,3,1970-01-01,Friday,Jan
3,0.00,4,1970-01-01,Friday,Jan
4,0.00,5,1970-01-01,Friday,Jan


### 1.2 Reconstruct the calendar and interval clock

Two cleanups make the data analysis-ready:

- The published `Date` field is a placeholder, but `Day` and `Month` carry the true 1999 calendar. We rebuild a real date index by walking the 365 days in order (1999-01-01 was a **Friday**, which matches the first day's label — a good integrity check).
- The interval index `u` (1–240) is converted to a wall-clock time. Interval 1 = 00:00–00:06, interval 2 = 00:06–00:12, and so on. We label each interval by its start time and also roll up to the **30-minute** grain that most WFM planning uses.


In [3]:
df = raw.copy()

# --- rebuild the true 1999 date index ---
# 240 intervals per day, days appear in calendar order.
n_days = df["Date"].nunique() if df["Date"].nunique() > 1 else len(df) // 240
day_block = np.repeat(np.arange(len(df) // 240), 240)
real_dates = pd.to_datetime("1999-01-01") + pd.to_timedelta(day_block, unit="D")
df["date"] = real_dates.values
df["dow"] = df["date"].dt.day_name()

# integrity check: dataset's own Day label should match the reconstructed weekday
check = (df["Day"].str.strip().values == df["dow"].values).mean()
print(f"Calendar reconstruction matches dataset Day label for {check:0.1%} of rows")

# --- interval clock ---
df["min_of_day"] = (df["u"] - 1) * 6
df["interval_start"] = pd.to_timedelta(df["min_of_day"], unit="m")
df["hh"] = df["min_of_day"] // 60
df["interval_30"] = (df["min_of_day"] // 30) * 30   # 30-min planning grain (minutes from midnight)

def hhmm(minutes):
    return f"{int(minutes)//60:02d}:{int(minutes)%60:02d}"

df["interval_label"] = df["min_of_day"].apply(hhmm)
df["interval_30_label"] = df["interval_30"].apply(hhmm)
df["is_weekend"] = df["dow"].isin(["Friday", "Saturday"])   # Israeli business week: Sun-Thu
df[["date", "dow", "u", "interval_label", "interval_30_label", "calls", "is_weekend"]].head()


Calendar reconstruction matches dataset Day label for 100.0% of rows


,date,dow,u,interval_label,interval_30_label,calls,is_weekend
0,1999-01-01,Friday,1,00:00,00:00,2.00,True
1,1999-01-01,Friday,2,00:06,00:00,0.00,True
2,1999-01-01,Friday,3,00:12,00:00,0.00,True
3,1999-01-01,Friday,4,00:18,00:00,0.00,True
4,1999-01-01,Friday,5,00:24,00:00,0.00,True


## 2. Exploratory Analysis — the shape of demand

Before forecasting, a WFM analyst characterizes the two seasonalities that drive every staffing plan: **day-of-week** and **time-of-day**.


In [4]:
# total calls per calendar day
daily = df.groupby(["date", "dow"], as_index=False)["calls"].sum()
print(f"Total calls in 1999: {df['calls'].sum():,.0f}")
print(f"Average calls per day: {daily['calls'].mean():,.0f}")
print(f"Busiest single day: {daily['calls'].max():,.0f}   Quietest: {daily['calls'].min():,.0f}")

order = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]
by_dow = daily.groupby("dow")["calls"].mean().reindex(order)

fig, ax = plt.subplots()
colors = ["#1E40AF" if d not in ["Friday","Saturday"] else "#94A3B8" for d in order]
ax.bar(by_dow.index, by_dow.values, color=colors)
ax.set_title("Average Daily Call Volume by Day of Week (1999)")
ax.set_ylabel("Avg calls / day")
ax.yaxis.set_major_formatter(mtick.StrMethodFormatter("{x:,.0f}"))
plt.xticks(rotation=20)
plt.tight_layout(); plt.show()


Total calls in 1999: 445,414
Average calls per day: 1,220
Busiest single day: 2,589   Quietest: 14


The business week is clearly **Sunday–Thursday**; Friday and Saturday are the weekend. This is the single most important pattern in the data — knowing only whether a day is a workday or a weekend explains the great majority of daily-volume variance. We carry the workday/weekend split through the rest of the analysis.


In [5]:
# intraday profile: average calls per 30-min interval, workdays only
workdays = df[~df["is_weekend"]]
intraday = (workdays.groupby("interval_30")["calls"]
            .mean()            # mean per 6-min row within each 30-min bucket-start... fix below
           )

# Correct 30-min volume: sum the 6-min rows into 30-min buckets per day, then average across days
per_day_30 = (workdays.groupby(["date", "interval_30"])["calls"].sum().reset_index())
profile_30 = per_day_30.groupby("interval_30")["calls"].mean()
profile_30.index = [hhmm(m) for m in profile_30.index]

fig, ax = plt.subplots(figsize=(14,5))
ax.plot(profile_30.index, profile_30.values, color="#1E40AF", lw=2, marker="o", ms=3)
ax.fill_between(range(len(profile_30)), profile_30.values, alpha=0.12, color="#1E40AF")
ax.set_title("Average Intraday Call Profile — 30-min intervals (Workdays, Sun-Thu)")
ax.set_ylabel("Avg calls per 30-min interval")
ax.set_xlabel("Interval start (HH:MM)")
step = 4
ax.set_xticks(range(0, len(profile_30), step))
ax.set_xticklabels(profile_30.index[::step], rotation=60, ha="right")
plt.tight_layout(); plt.show()


The intraday curve shows the classic contact-center signature: an overnight trough, a sharp morning ramp from ~07:00, a mid-morning peak, a midday softening, a second afternoon shoulder, then an evening decline. A flat, average-staffing plan would badly over-staff the night and under-staff the morning peak — which is exactly why WFM forecasts and schedules at the **interval** level rather than the daily level.


## 3. Interval-Level Volume Forecast

**Method.** A practical, transparent WFM forecast separates the two components an analyst reasons about directly:

- a **day-level volume** expectation (how many contacts tomorrow), and
- an **intraday distribution profile** (what fraction of the day's volume lands in each interval).

We build a **profile-ratio (a.k.a. percent-of-day) forecast**, the workhorse method on most WFM floors:

1. Compute each workday's interval *share* of that day's total.
2. Average those shares by `(day-of-week, interval)` to get a stable intraday **distribution profile**.
3. Forecast the **day total** for a target day (here: a seasonal-naive / day-of-week average baseline).
4. Multiply: `interval_forecast = day_total_forecast × interval_share`.

This decomposition is intentionally interpretable — every number a planner sees traces back to either "expected daily volume" or "shape of the day," both of which are quantities operations leaders can challenge and adjust. We validate it with a holdout.


In [6]:
# Work on 30-min intervals, workdays only, for a clean staffing demonstration
vol = (df.groupby(["date","dow","is_weekend","interval_30"])["calls"].sum().reset_index())
vol = vol[~vol["is_weekend"]].copy()

# day totals
day_tot = vol.groupby("date")["calls"].transform("sum")
vol["interval_share"] = np.where(day_tot > 0, vol["calls"] / day_tot, 0)

# ---- train / test split: last 4 workweeks as holdout ----
all_dates = np.sort(vol["date"].unique())
split_date = all_dates[-20]      # ~last 4 Sun-Thu weeks as test
train = vol[vol["date"] <  split_date]
test  = vol[vol["date"] >= split_date]
print(f"Train days: {train['date'].nunique()}  |  Test days: {test['date'].nunique()}")

# ---- intraday distribution profile by (dow, interval) from TRAIN ----
profile = (train.groupby(["dow","interval_30"])["interval_share"].mean().reset_index()
           .rename(columns={"interval_share":"profile_share"}))
# renormalize each dow profile to sum to 1
profile["profile_share"] /= profile.groupby("dow")["profile_share"].transform("sum")

# ---- day-total forecast: day-of-week mean of daily totals from TRAIN ----
train_day_totals = train.groupby(["date","dow"])["calls"].sum().reset_index()
dow_total_fc = train_day_totals.groupby("dow")["calls"].mean().rename("day_total_fc")
print("\nForecast daily total by weekday (train):")
print(dow_total_fc.round(0).to_string())


Train days: 240  |  Test days: 20

Forecast daily total by weekday (train):
dow
Monday      1,542.00
Sunday      1,616.00
Thursday    1,506.00
Tuesday     1,576.00
Wednesday   1,480.00


In [7]:
# ---- build interval forecast for the test window ----
fc = test.merge(profile, on=["dow","interval_30"], how="left")
fc = fc.merge(dow_total_fc, on="dow", how="left")
fc["forecast"] = fc["day_total_fc"] * fc["profile_share"]

# accuracy metrics at the interval level
actual = fc["calls"].values
pred   = fc["forecast"].values
mae  = np.mean(np.abs(actual - pred))
rmse = np.sqrt(np.mean((actual - pred)**2))
mask = actual > 0
mape = np.mean(np.abs((actual[mask]-pred[mask]) / actual[mask])) * 100
wape = np.sum(np.abs(actual - pred)) / np.sum(actual) * 100   # volume-weighted; the WFM-standard error metric

print(f"Interval-level forecast accuracy (holdout, 30-min intervals)")
print(f"  MAE  : {mae:6.2f} calls/interval")
print(f"  RMSE : {rmse:6.2f} calls/interval")
print(f"  MAPE : {mape:6.1f}%   (unstable on low-volume intervals)")
print(f"  WAPE : {wape:6.1f}%   (volume-weighted -- the metric WFM teams track)")


Interval-level forecast accuracy (holdout, 30-min intervals)
  MAE  :   7.84 calls/interval
  RMSE :  12.47 calls/interval
  MAPE :   28.5%   (unstable on low-volume intervals)
  WAPE :   21.2%   (volume-weighted -- the metric WFM teams track)


**Why WAPE, not MAPE.** MAPE blows up on low-volume overnight intervals (dividing by 1 or 2 calls), so it overstates error for a staffing context where those intervals barely matter. WAPE (weighted absolute percent error) weights each interval by its volume, which is what actually drives headcount. Reporting both is honest; planning on WAPE is standard.


In [8]:
# visualize forecast vs actual for a representative test day
sample_day = np.sort(test["date"].unique())[5]
day_fc = fc[fc["date"] == sample_day].sort_values("interval_30")
xlab = [hhmm(m) for m in day_fc["interval_30"]]

fig, ax = plt.subplots(figsize=(14,5))
ax.plot(xlab, day_fc["calls"].values,    color="#0F172A", lw=2, marker="o", ms=3, label="Actual")
ax.plot(xlab, day_fc["forecast"].values, color="#DC2626", lw=2, ls="--", marker="s", ms=3, label="Forecast")
ax.set_title(f"Interval Forecast vs Actual — {pd.Timestamp(sample_day).strftime('%A, %d %b %Y')}")
ax.set_ylabel("Calls per 30-min interval"); ax.set_xlabel("Interval start (HH:MM)")
ax.legend()
step = 4
ax.set_xticks(range(0, len(xlab), step)); ax.set_xticklabels(xlab[::step], rotation=60, ha="right")
plt.tight_layout(); plt.show()


The profile-ratio forecast tracks the intraday shape well. In production this baseline is where tuning begins — adding trend, holiday/event adjustments, and intraday re-forecasting — but a clean, explainable baseline is the right foundation and a fair benchmark for any fancier model.


## 4. From Calls to Workload — Average Handle Time

Volume alone doesn't size a team; **workload** does. Workload per interval = `forecast calls × Average Handle Time (AHT)`. AHT (talk + after-call work) is not in this arrivals dataset, so we introduce it as an explicit, labeled WFM assumption.

> **SYNTHETIC WFM ASSUMPTION:** `AHT = 240 seconds` (4.0 min) average handle time, a realistic retail-bank figure. This is a planning input a real WFM team pulls from the ACD/telephony platform; here it is stated openly rather than derived from the data.


In [9]:
AHT_SECONDS = 240          # SYNTHETIC ASSUMPTION: average handle time (talk + ACW)
INTERVAL_SECONDS = 30 * 60 # 30-minute planning interval

# Build the forward staffing plan: use the dow profile + dow day-total forecast
# to produce an expected-volume plan for one representative workday of each weekday.
plan = (profile.merge(dow_total_fc, on="dow", how="left"))
plan["forecast_calls"] = plan["day_total_fc"] * plan["profile_share"]
plan["workload_erlangs"] = plan["forecast_calls"] * AHT_SECONDS / INTERVAL_SECONDS  # offered load A
plan = plan.sort_values(["dow","interval_30"]).reset_index(drop=True)
plan[["dow","interval_30","forecast_calls","workload_erlangs"]].head()


,dow,interval_30,forecast_calls,workload_erlangs
0,Monday,0,3.05,0.41
1,Monday,30,2.28,0.30
2,Monday,60,1.22,0.16
3,Monday,90,0.99,0.13
4,Monday,120,0.48,0.06


`workload_erlangs` is the **offered load (A)** — the average number of calls in progress during the interval if there were no waiting. It is the input to Erlang C. (One Erlang = one call-hour of work per hour.)


## 5. Erlang C — Staffing for a Service Level

Contact centers don't staff to raw workload; they staff to a **service-level target** while accounting for the randomness of arrivals. The standard model is **Erlang C**, which—given offered load `A` and `N` agents—returns the probability a call must wait, and from that the **service level** (the share of calls answered within a target time).

> **SYNTHETIC WFM ASSUMPTION — Service-Level Target:** **80/20** — answer 80% of calls within 20 seconds. The most common contact-center SLA.

Erlang C assumptions (and their real-world caveats):
- Calls arrive as a Poisson process; handle times are exponential.
- Infinite queue patience (no abandonment). Real centers abandon, so Erlang C is mildly **conservative** (slightly over-staffs). Erlang A/X refine this; Erlang C remains the planning default for its simplicity and transparency.


In [10]:
def erlang_c_prob_wait(A, N):
    """Probability a call waits (Erlang C), offered load A (erlangs), N agents."""
    if N <= A:
        return 1.0  # unstable: arrivals exceed capacity
    # sum_{k=0}^{N-1} A^k / k!
    s = sum(A**k / factorial(k) for k in range(N))
    last = A**N / factorial(N) * (N / (N - A))
    return last / (s + last)

def service_level(A, N, target_seconds, aht_seconds):
    """Fraction of calls answered within target_seconds."""
    if N <= A:
        return 0.0
    pw = erlang_c_prob_wait(A, N)
    return 1 - pw * np.exp(-(N - A) * target_seconds / aht_seconds)

def occupancy(A, N):
    return A / N if N > 0 else 1.0

def agents_required(A, target_sl, target_seconds, aht_seconds, max_occ=0.90):
    """Smallest N meeting the SL target AND an occupancy ceiling."""
    if A <= 0:
        return 0
    N = max(1, int(np.floor(A)) + 1)
    while True:
        sl = service_level(A, N, target_seconds, aht_seconds)
        occ = occupancy(A, N)
        if sl >= target_sl and occ <= max_occ:
            return N
        N += 1
        if N > 1000:
            return N

# quick sanity demo
demoA = 5.0
for n in range(5, 11):
    print(f"A={demoA}  N={n:2d}  P(wait)={erlang_c_prob_wait(demoA,n):.3f}  "
          f"SL(20s)={service_level(demoA,n,20,AHT_SECONDS):.1%}  occ={occupancy(demoA,n):.1%}")


A=5.0  N= 5  P(wait)=1.000  SL(20s)=0.0%  occ=100.0%
A=5.0  N= 6  P(wait)=0.588  SL(20s)=45.9%  occ=83.3%
A=5.0  N= 7  P(wait)=0.324  SL(20s)=72.6%  occ=71.4%
A=5.0  N= 8  P(wait)=0.167  SL(20s)=87.0%  occ=62.5%
A=5.0  N= 9  P(wait)=0.081  SL(20s)=94.2%  occ=55.6%
A=5.0  N=10  P(wait)=0.036  SL(20s)=97.6%  occ=50.0%


> **SYNTHETIC WFM ASSUMPTION — Occupancy ceiling 90%.** We require occupancy ≤ 90% so the model never plans agents to be busy 100% of the time — sustained high occupancy drives burnout and attrition. This is a humane, standard planning guardrail.


In [11]:
TARGET_SL  = 0.80
TARGET_SEC = 20
MAX_OCC    = 0.90

plan["agents_required"] = plan["workload_erlangs"].apply(
    lambda A: agents_required(A, TARGET_SL, TARGET_SEC, AHT_SECONDS, MAX_OCC))
plan["proj_service_level"] = plan.apply(
    lambda r: service_level(r["workload_erlangs"], r["agents_required"], TARGET_SEC, AHT_SECONDS), axis=1)
plan["proj_occupancy"] = plan.apply(
    lambda r: occupancy(r["workload_erlangs"], r["agents_required"]), axis=1)

plan[["dow","interval_30","forecast_calls","workload_erlangs",
      "agents_required","proj_service_level","proj_occupancy"]].head(12)


,dow,interval_30,forecast_calls,workload_erlangs,agents_required,proj_service_level,proj_occupancy
0,Monday,0,3.05,0.41,2,0.94,0.20
1,Monday,30,2.28,0.30,2,0.97,0.15
2,Monday,60,1.22,0.16,1,0.85,0.16
3,Monday,90,0.99,0.13,1,0.88,0.13
4,Monday,120,0.48,0.06,1,0.94,0.06
5,Monday,150,0.58,0.08,1,0.93,0.08
6,Monday,180,0.24,0.03,1,0.97,0.03
7,Monday,210,0.24,0.03,1,0.97,0.03
8,Monday,240,0.25,0.03,1,0.97,0.03
9,Monday,270,0.16,0.02,1,0.98,0.02


## 6. Shrinkage — From Agents-on-Phone to Scheduled FTEs

`agents_required` is the number of bodies that must be **on the phone** in each interval. But scheduled staff are not on the phone 100% of their paid time. **Shrinkage** is the portion of paid time lost to everything that isn't handling contacts: breaks, lunch, training, meetings, coaching, paid time off, sick, and general unproductive time.

The required *scheduled* headcount is:

$$\text{Scheduled FTEs} = \frac{\text{Agents required on phone}}{1 - \text{Shrinkage}}$$

> **SYNTHETIC WFM ASSUMPTION — Shrinkage build-up.** Decomposed transparently so each component can be challenged:

| Shrinkage component | Rate |
|---|---|
| Breaks & lunch | 12% |
| Training & meetings | 6% |
| Coaching / 1:1s | 3% |
| PTO / sick / absence | 8% |
| Unproductive / shoulder time | 5% |
| **Total shrinkage** | **34%** |

Owning the shrinkage build-up was a core part of my MCI intraday role — coding schedule exceptions (training, meetings, approved offline time) so adherence and shrinkage reported accurately at team and center level. The component view below is exactly how that reporting is structured.


In [12]:
SHRINKAGE = {
    "Breaks & lunch":            0.12,
    "Training & meetings":       0.06,
    "Coaching / 1:1s":           0.03,
    "PTO / sick / absence":      0.08,
    "Unproductive / shoulder":   0.05,
}
TOTAL_SHRINKAGE = sum(SHRINKAGE.values())
print(f"Total shrinkage: {TOTAL_SHRINKAGE:.0%}")

# shrinkage uplift
plan["scheduled_fte"] = np.ceil(plan["agents_required"] / (1 - TOTAL_SHRINKAGE)).astype(int)

# visualize the build-up
fig, ax = plt.subplots(figsize=(9,4))
comp = pd.Series(SHRINKAGE).sort_values()
ax.barh(comp.index, comp.values*100, color="#1E40AF")
ax.xaxis.set_major_formatter(mtick.PercentFormatter())
ax.set_title(f"Shrinkage Build-Up (Total {TOTAL_SHRINKAGE:.0%}) — SYNTHETIC assumption")
for i,v in enumerate(comp.values):
    ax.text(v*100+0.3, i, f"{v:.0%}", va="center", fontsize=10)
plt.tight_layout(); plt.show()

plan[["dow","interval_30","agents_required","scheduled_fte"]].head(12)


Total shrinkage: 34%


,dow,interval_30,agents_required,scheduled_fte
0,Monday,0,2,4
1,Monday,30,2,4
2,Monday,60,1,2
3,Monday,90,1,2
4,Monday,120,1,2
5,Monday,150,1,2
6,Monday,180,1,2
7,Monday,210,1,2
8,Monday,240,1,2
9,Monday,270,1,2


## 7. Staffing Gap Analysis

The payoff of the whole pipeline: compare the **required** scheduled FTEs against what is **actually scheduled**, interval by interval, and flag the gaps. Over-staffing wastes money; under-staffing misses SLA and burns out the team.

> **SYNTHETIC SCHEDULE:** Real centers schedule in shift blocks, not to the per-interval ideal, so coverage is "lumpy." We emulate this by scheduling a smoothed, shift-shaped headcount: a flat core staff across open hours plus a daytime bump, deliberately *not* matching the curve. This produces the realistic over/under pattern a WFM analyst hunts for.


In [13]:
# Focus on one representative weekday for the gap demonstration
demo_dow = "Tuesday"
d = plan[plan["dow"] == demo_dow].sort_values("interval_30").reset_index(drop=True)

# --- synthetic shift-based schedule (lumpy, not curve-matched) ---
# Built from overlapping shift blocks so total headcount is roughly right
# but the *distribution* lags the forecast curve -- the realistic failure mode.
m = d["interval_30"]
early = ((m >= 7*60)  & (m < 16*60)).astype(int) * 12   # early shift 07:00-16:00
mid   = ((m >= 10*60) & (m < 19*60)).astype(int) * 8    # mid shift   10:00-19:00 (lands AFTER the 10:00 peak)
late  = ((m >= 15*60) & (m < 24*60)).astype(int) * 7    # late shift  15:00-24:00
d["scheduled_actual"] = (early + mid + late)

d["gap"] = d["scheduled_actual"] - d["scheduled_fte"]   # +ve = overstaffed, -ve = understaffed

# recompute the service level you'd actually deliver with the lumpy schedule
def delivered_sl(row):
    on_phone = row["scheduled_actual"] * (1 - TOTAL_SHRINKAGE)
    A = row["workload_erlangs"]
    if A <= 0:
        return 1.0
    return service_level(A, max(1, int(np.floor(on_phone))), TARGET_SEC, AHT_SECONDS)

d["delivered_sl"] = d.apply(delivered_sl, axis=1)
d[["interval_30","forecast_calls","scheduled_fte","scheduled_actual","gap","delivered_sl"]].iloc[14:40:2]


,interval_30,forecast_calls,scheduled_fte,scheduled_actual,gap,delivered_sl
14,420,16.54,7,12,5,0.99
16,480,36.49,13,12,-1,0.75
18,540,56.03,17,12,-5,0.00
20,600,73.34,20,20,0,0.81
22,660,59.84,17,20,3,0.95
24,720,54.91,17,20,3,0.97
26,780,51.27,16,20,4,0.98
28,840,65.87,19,20,1,0.90
30,900,60.98,19,27,8,1.00
32,960,61.85,19,15,-4,0.31


In [14]:
xlab = [hhmm(m) for m in d["interval_30"]]
fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(14, 9), sharex=True,
                               gridspec_kw={"height_ratios":[2,1]})

# top: required vs scheduled
ax1.plot(xlab, d["scheduled_fte"], color="#1E40AF", lw=2, marker="o", ms=3, label="Required FTEs (forecast)")
ax1.plot(xlab, d["scheduled_actual"], color="#059669", lw=2, ls="--", marker="s", ms=3, label="Scheduled FTEs (actual)")
ax1.fill_between(range(len(d)), d["scheduled_fte"], d["scheduled_actual"],
                 where=(d["scheduled_actual"] >= d["scheduled_fte"]), color="#10B981", alpha=0.18, label="Overstaffed")
ax1.fill_between(range(len(d)), d["scheduled_fte"], d["scheduled_actual"],
                 where=(d["scheduled_actual"] <  d["scheduled_fte"]), color="#EF4444", alpha=0.20, label="Understaffed")
ax1.set_title(f"Staffing Gap Analysis — {demo_dow} (Required vs Scheduled, 30-min intervals)")
ax1.set_ylabel("FTEs"); ax1.legend(loc="upper right", ncol=2, fontsize=9)

# closed hours (no one scheduled) are shaded and excluded from service level
closed = d["scheduled_actual"] == 0
for ax in (ax1, ax2):
    ax.axvspan(-0.5, closed.values.argmin() - 0.5, color="#E5E7EB", alpha=0.6, lw=0)
ax1.text(closed.values.argmin()/2, ax1.get_ylim()[1]*0.9, "Closed\n(unscheduled)", ha="center", fontsize=9, color="#6B7280")

# bottom: delivered service level (staffed intervals only)
ax2.plot(xlab, d["delivered_sl"].where(~closed)*100, color="#7C3AED", lw=2, marker="o", ms=3)
ax2.axhline(TARGET_SL*100, color="#DC2626", ls=":", lw=1.5, label=f"SL target {TARGET_SL:.0%}")
ax2.set_ylabel("Delivered SL (%)"); ax2.set_ylim(0, 105)
ax2.yaxis.set_major_formatter(mtick.PercentFormatter())
ax2.legend(loc="lower right", fontsize=9)

step = 4
ax2.set_xticks(range(0, len(xlab), step)); ax2.set_xticklabels(xlab[::step], rotation=60, ha="right")
ax2.set_xlabel("Interval start (HH:MM)")
plt.tight_layout()
plt.savefig('staffing_gap_chart.png', dpi=150, bbox_inches='tight')
plt.show()


Read this the way a WFM analyst would on a Monday morning:

- **Red bands (understaffed)** — typically the morning ramp and peak, where a lumpy shift schedule lags the forecast curve. These are the intervals where SLA breaks and customers wait. The fix is shift-start staggering or a part-time tranche landing at the peak.
- **Green bands (overstaffed)** — usually the late evening and shoulders, where the flat core over-covers low volume. These are cost-reduction or offline-work (training, coaching) opportunities.
- **Bottom panel** — the service level you'd actually deliver. Where the purple line dips below the red SLA line, the schedule fails the 80/20 target *regardless* of the daily headcount being "enough" — proving the WFM point that **total headcount can be right while the intraday distribution is wrong.**


In [15]:
# summary scorecard for the day
open_intervals = d[d["scheduled_actual"] > 0]
under = (open_intervals["gap"] < 0).sum()
over  = (open_intervals["gap"] > 0).sum()
sla_miss = (open_intervals["delivered_sl"] < TARGET_SL).sum()
agg_sl = np.average(open_intervals["delivered_sl"], weights=open_intervals["forecast_calls"])

print(f"=== {demo_dow} WFM Scorecard ===")
print(f"Staffed intervals          : {len(open_intervals)}")
print(f"Understaffed intervals     : {under}  ({under/len(open_intervals):.0%})")
print(f"Overstaffed intervals      : {over}  ({over/len(open_intervals):.0%})")
print(f"Intervals missing SLA      : {sla_miss}  ({sla_miss/len(open_intervals):.0%})")
print(f"Volume-weighted day SL     : {agg_sl:.1%}   (target {TARGET_SL:.0%})")
print(f"Total scheduled FTE-intervals: {int(open_intervals['scheduled_actual'].sum())}")
print(f"Total required  FTE-intervals: {int(open_intervals['scheduled_fte'].sum())}")


=== Tuesday WFM Scorecard ===
Staffed intervals          : 34
Understaffed intervals     : 18  (53%)
Overstaffed intervals      : 15  (44%)
Intervals missing SLA      : 18  (53%)
Volume-weighted day SL     : 62.4%   (target 80%)
Total scheduled FTE-intervals: 486
Total required  FTE-intervals: 494


---
## Module 5: Workforce Optimization Engine

The gap analysis in Section 7 *described* the problem — 53% of intervals understaffed, service level 62.4% against an 80% target. This module moves from description to **recommendation**.

The optimization engine answers the operational question a WFM manager asks every morning:

> *"Given the gaps I can see, what specific moves would recover the most service level — and at what cost?"*

**Approach:** The engine identifies overstaffed intervals and recommends agent redistribution to understaffed intervals, recalculates delivered service level after each move, and quantifies the improvement — without adding headcount.

This is the core of intraday WFM: **same people, better distribution.**


In [16]:
# -- Module 5: Optimization Engine ---------------------------------------------
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick

# Work from the Tuesday gap analysis dataframe (d) built in Section 7
opt = d.copy()

# ── Step 1: Identify redistribution candidates ─────────────────────────────
# Overstaffed intervals: scheduled > required (surplus agents available)
# Understaffed intervals: scheduled < required (agents needed)
# Only operate during staffed hours (scheduled_actual > 0)

opt['surplus']  = (opt['scheduled_actual'] - opt['scheduled_fte']).clip(lower=0)
opt['deficit']  = (opt['scheduled_fte']   - opt['scheduled_actual']).clip(lower=0)

total_surplus = opt['surplus'].sum()
total_deficit = opt['deficit'].sum()

print(f'Total surplus agent-intervals available: {total_surplus:.0f}')
print(f'Total deficit agent-intervals to fill:   {total_deficit:.0f}')
print(f'Redistribution coverage potential:       {min(total_surplus/total_deficit,1):.0%}')
print()


Total surplus agent-intervals available: 51
Total deficit agent-intervals to fill:   95
Redistribution coverage potential:       54%



In [17]:
# -- Step 2: Greedy redistribution algorithm -----------------------------------
# Priority: fill deficits in order of call volume (highest-volume understaffed
# intervals get agents first — maximizes service level recovery per move)

opt_schedule = opt['scheduled_actual'].copy()
moves = []  # log each redistribution move

# Build priority queues
# Understaffed intervals sorted by call volume descending (highest impact first)
understaffed_idx = opt[opt['deficit'] > 0].sort_values('forecast_calls', ascending=False).index.tolist()
# Overstaffed intervals sorted by surplus descending (most available first)
overstaffed_idx  = opt[opt['surplus'] > 0].sort_values('surplus', ascending=False).index.tolist()

surplus_remaining = opt['surplus'].copy()
deficit_remaining = opt['deficit'].copy()

for dest in understaffed_idx:
    needed = deficit_remaining[dest]
    if needed <= 0:
        continue
    for src in overstaffed_idx:
        available = surplus_remaining[src]
        if available <= 0:
            continue
        # How many agents can we move?
        transfer = min(needed, available)
        opt_schedule[dest] += transfer
        opt_schedule[src]  -= transfer
        surplus_remaining[src]  -= transfer
        deficit_remaining[dest] -= transfer
        moves.append({
            'from_interval': hhmm(opt.loc[src,  'interval_30']),
            'to_interval':   hhmm(opt.loc[dest, 'interval_30']),
            'agents_moved':  round(transfer, 1),
            'dest_volume':   int(opt.loc[dest, 'forecast_calls']),
        })
        needed -= transfer
        if needed <= 0:
            break

opt['optimized_schedule'] = opt_schedule
opt['optimized_gap'] = opt['optimized_schedule'] - opt['scheduled_fte']

# Recompute delivered SL under optimized schedule
def delivered_sl_n(row, col):
    on_phone = row[col] * (1 - TOTAL_SHRINKAGE)
    A = row['workload_erlangs']
    if A <= 0: return 1.0
    return service_level(A, max(1, int(np.floor(on_phone))), TARGET_SEC, AHT_SECONDS)

opt['optimized_sl'] = opt.apply(lambda r: delivered_sl_n(r, 'optimized_schedule'), axis=1)

# Top 10 recommended moves
moves_df = pd.DataFrame(moves).head(10)
print('Top redistribution recommendations:')
print(moves_df.to_string(index=False))
print(f'\nTotal moves generated: {len(moves)}')


Top redistribution recommendations:
from_interval to_interval  agents_moved  dest_volume
        15:30       09:30             7           63
        15:30       16:00             3           61
        15:00       16:00             1           61
        15:00       16:30             2           59
        15:00       09:00             5           56
        07:00       08:30             4           51
        07:00       17:00             1           50
        12:30       17:30             1           48
        12:30       08:00             1           36
        12:30       19:00             2           32

Total moves generated: 24


In [18]:
# -- Step 3: Quantify service level recovery -----------------------------------

open_mask = opt['scheduled_actual'] > 0
op = opt[open_mask]

# Before optimization
sl_before = np.average(op['delivered_sl'],   weights=op['forecast_calls'])
under_before = (op['gap'] < 0).sum()
over_before  = (op['gap'] > 0).sum()

# After optimization
sl_after  = np.average(op['optimized_sl'],  weights=op['forecast_calls'])
under_after  = (op['optimized_gap'] < 0).sum()
over_after   = (op['optimized_gap'] > 0).sum()

sl_recovery = sl_after - sl_before

print('=' * 52)
print('  OPTIMIZATION ENGINE — RESULTS')
print('=' * 52)
print(f'  {"Metric":<30} {"Before":>8}  {"After":>8}')
print(f'  {"-"*48}')
print(f'  {"Volume-weighted SL":<30} {sl_before:>7.1%}  {sl_after:>7.1%}')
print(f'  {"Understaffed intervals":<30} {under_before:>8}  {under_after:>8}')
print(f'  {"Overstaffed intervals":<30} {over_before:>8}  {over_after:>8}')
print(f'  {"SL gap vs 80% target":<30} {TARGET_SL-sl_before:>7.1%}  {TARGET_SL-sl_after:>7.1%}')
print(f'  {"-"*48}')
print(f'  {"SL Recovery":<30} {sl_recovery*100:>+6.1f} pp')
print(f'  {"Headcount added":<30} {0:>8}')
print('=' * 52)
print()
print(f'  Key insight: {sl_recovery*100:+.1f} percentage-point service level improvement')
print(f'  achieved by redistribution alone — zero additional FTEs.')


  OPTIMIZATION ENGINE — RESULTS
  Metric                           Before     After
  ------------------------------------------------
  Volume-weighted SL               62.4%    82.5%
  Understaffed intervals               18         4
  Overstaffed intervals                15         0
  SL gap vs 80% target             17.6%    -2.5%
  ------------------------------------------------
  SL Recovery                     +20.2 pp
  Headcount added                       0

  Key insight: +20.2 percentage-point service level improvement
  achieved by redistribution alone — zero additional FTEs.


In [19]:
# -- Step 4: Before vs After visualization ------------------------------------

xlab = [hhmm(m) for m in opt['interval_30']]

fig, axes = plt.subplots(3, 1, figsize=(14, 12), sharex=True,
                         gridspec_kw={"height_ratios": [2, 2, 1.2]})
fig.suptitle('Optimization Engine — Before vs After Redistribution', fontsize=14, fontweight='bold')

# ── Panel 1: Before ───────────────────────────────────────────────────────
ax = axes[0]
ax.plot(xlab, opt['scheduled_fte'],    color='#1E40AF', lw=2, label='Required FTEs')
ax.plot(xlab, opt['scheduled_actual'], color='#DC2626', lw=2, ls='--', label='Scheduled (before)')
ax.fill_between(range(len(opt)), opt['scheduled_fte'], opt['scheduled_actual'],
    where=(opt['scheduled_actual'] < opt['scheduled_fte']), color='#EF4444', alpha=0.20, label='Understaffed')
ax.fill_between(range(len(opt)), opt['scheduled_fte'], opt['scheduled_actual'],
    where=(opt['scheduled_actual'] >= opt['scheduled_fte']), color='#10B981', alpha=0.15, label='Overstaffed')
ax.set_title('Before Optimization', fontweight='bold', fontsize=11)
ax.set_ylabel('FTEs')
ax.legend(loc='upper right', ncol=2, fontsize=9)
ax.annotate(f'SL: {sl_before:.1%}', xy=(0.02, 0.88), xycoords='axes fraction',
            fontsize=11, color='#DC2626', fontweight='bold')

# ── Panel 2: After ────────────────────────────────────────────────────────
ax = axes[1]
ax.plot(xlab, opt['scheduled_fte'],       color='#1E40AF', lw=2, label='Required FTEs')
ax.plot(xlab, opt['optimized_schedule'],  color='#059669', lw=2, ls='--', label='Scheduled (optimized)')
ax.fill_between(range(len(opt)), opt['scheduled_fte'], opt['optimized_schedule'],
    where=(opt['optimized_schedule'] < opt['scheduled_fte']), color='#EF4444', alpha=0.20, label='Understaffed')
ax.fill_between(range(len(opt)), opt['scheduled_fte'], opt['optimized_schedule'],
    where=(opt['optimized_schedule'] >= opt['scheduled_fte']), color='#10B981', alpha=0.15, label='Overstaffed')
ax.set_title('After Optimization (Redistribution Only)', fontweight='bold', fontsize=11)
ax.set_ylabel('FTEs')
ax.legend(loc='upper right', ncol=2, fontsize=9)
ax.annotate(f'SL: {sl_after:.1%}', xy=(0.02, 0.88), xycoords='axes fraction',
            fontsize=11, color='#059669', fontweight='bold')

# ── Panel 3: SL comparison ────────────────────────────────────────────────
ax = axes[2]
closed = opt['scheduled_actual'] == 0
ax.plot(xlab, opt['delivered_sl'].where(~closed)  * 100, color='#DC2626', lw=2, ls='--', label=f'Before ({sl_before:.1%})')
ax.plot(xlab, opt['optimized_sl'].where(~closed)  * 100, color='#059669', lw=2,          label=f'After  ({sl_after:.1%})')
ax.axhline(TARGET_SL * 100, color='#1E40AF', ls=':', lw=1.5, label=f'Target {TARGET_SL:.0%}')
ax.set_ylabel('Service Level (%)')
ax.set_ylim(0, 105)
ax.yaxis.set_major_formatter(mtick.PercentFormatter())
ax.set_title('Delivered Service Level — Before vs After (staffed intervals)', fontweight='bold', fontsize=11)
ax.legend(loc='lower right', ncol=3, fontsize=9)

step = 4
ax.set_xticks(range(0, len(xlab), step))
ax.set_xticklabels(xlab[::step], rotation=60, ha='right')
ax.set_xlabel('Interval start (HH:MM)')

plt.tight_layout()
plt.savefig('optimization_before_after.png', dpi=150, bbox_inches='tight')
plt.show()
print('Chart saved: optimization_before_after.png')


Chart saved: optimization_before_after.png


**What the optimization engine demonstrates:**

- Redistribution recovers meaningful service level improvement with **zero additional headcount**
- The greedy algorithm prioritizes high-volume intervals first — maximizing SL recovery per agent moved
- Residual gap after redistribution represents the *irreducible* deficit: intervals where demand exceeds total available supply regardless of distribution
- In production, this logic runs intraday — WFM analysts use it to make real-time schedule adjustments, VTO offers, and agent recall decisions

> *The optimization engine answers the question gap analysis raises: not just "where are we short" but "what specifically should we do about it."*


---
## Module 6: Decision Lab — What-If Scenario Simulator

The optimization engine redistributes existing capacity. The Decision Lab answers a different class of question:

> *"If we change a planning assumption, how much does service level move?"*

This is the **what-if layer** WFM managers use to justify headcount decisions, negotiate shrinkage targets, and evaluate the ROI of AHT reduction initiatives. Each lever below is independently adjustable — the model recalculates delivered service level for each scenario.

**Levers modeled** (all changes to service level are reported in percentage points):
1. Redistribute existing labor (no new hires, total staffed agent-intervals held constant)
2. Add headcount: **+10%** of scheduled agent-intervals, the comparison most budget requests make
3. Add headcount: **+5 FTEs in every staffed interval** (+35% of scheduled agent-intervals, shown as an upper bound)
4. Reduce AHT 15%
5. Reduce shrinkage 4 points
6. Combined: redistribute + AHT 15% + shrinkage 4 points


In [20]:
# -- Module 6: Decision Lab ----------------------------------------------------

def simulate_scenario(label, schedule_override=None, aht_override=None,
                       shrinkage_override=None, volume_scale=1.0):
    """
    Simulate a what-if scenario and return volume-weighted service level.
    Defaults to Tuesday base case. Override any lever independently.
    """
    sim = d.copy()
    aht      = aht_override      if aht_override      is not None else AHT_SECONDS
    shrink   = shrinkage_override if shrinkage_override is not None else TOTAL_SHRINKAGE
    sched    = schedule_override  if schedule_override  is not None else sim['scheduled_actual']

    sim['sim_calls']    = sim['forecast_calls'] * volume_scale
    sim['sim_erlangs']  = (sim['sim_calls'] * aht) / (30 * 60)
    sim['sim_schedule'] = sched

    def sim_sl(row):
        on_phone = row['sim_schedule'] * (1 - shrink)
        A = row['sim_erlangs']
        if A <= 0: return 1.0
        return service_level(A, max(1, int(np.floor(on_phone))), TARGET_SEC, aht)

    open_mask = sim['sim_schedule'] > 0
    sim['sim_sl'] = sim.apply(sim_sl, axis=1)
    op = sim[open_mask]
    weighted_sl = np.average(op['sim_sl'], weights=op['sim_calls'])
    return {"Scenario": label, 'Service Level': weighted_sl}


# ── Define scenarios ──────────────────────────────────────────────────────

# Baseline
base_sl = simulate_scenario('Baseline (current schedule)')

# Lever 1: Redistribution only (use optimized schedule from Module 5)
s1 = simulate_scenario('Lever 1: Redistribute (no new hires)',
                        schedule_override=opt['optimized_schedule'])

# Lever 2: Add 10% headcount (rounded per interval) -- comparable to a typical budget ask
add_10 = (d['scheduled_actual'] * 1.10).round()
s2 = simulate_scenario('Lever 2: Add 10% headcount', schedule_override=add_10)

# Lever 2b: Add 5 FTEs in every staffed interval -- +35% of scheduled agent-intervals (upper bound)
add_fte = d['scheduled_actual'].copy()
add_fte[d['scheduled_actual'] > 0] += 5
s2b = simulate_scenario('Lever 2b: Add 5 FTEs/interval (+35%)', schedule_override=add_fte)
print(f"Scheduled agent-intervals: base {int(d['scheduled_actual'].sum())} | +10% lever {int(add_10.sum())} | +5 FTE lever {int(add_fte.sum())}")

# Lever 3: Reduce AHT by 15% (coaching / process improvement initiative)
s3 = simulate_scenario('Lever 3: Reduce AHT 15%',
                        aht_override=AHT_SECONDS * 0.85)

# Lever 4: Reduce shrinkage by 4 points (tighten break scheduling)
s4 = simulate_scenario('Lever 4: Reduce shrinkage 4pts',
                        shrinkage_override=TOTAL_SHRINKAGE - 0.04)

# Lever 5: Combined -- redistribution plus both efficiency levers
s5 = simulate_scenario('Lever 5: Redistribute + AHT 15% + Shrinkage 4pts',
                        schedule_override=opt['optimized_schedule'],
                        aht_override=AHT_SECONDS * 0.85,
                        shrinkage_override=TOTAL_SHRINKAGE - 0.04)

scenarios = pd.DataFrame([base_sl, s1, s2, s2b, s3, s4, s5])
scenarios['vs Baseline'] = scenarios['Service Level'] - scenarios.loc[0, 'Service Level']
scenarios['Hits 80% Target'] = scenarios['Service Level'] >= TARGET_SL

print('=' * 65)
print('  DECISION LAB — WHAT-IF SCENARIO RESULTS')
print('=' * 65)
for _, row in scenarios.iterrows():
    hit = '✓ TARGET MET' if row['Hits 80% Target'] else '✗ below target'
    delta = f"{row['vs Baseline']*100:+.1f} pp"
    print(f"  {row['Scenario'][:45]:<45} {row['Service Level']:>6.1%}  {delta:>7}  {hit}")
print('=' * 65)


Scheduled agent-intervals: base 486 | +10% lever 534 | +5 FTE lever 656
  DECISION LAB — WHAT-IF SCENARIO RESULTS
  Baseline (current schedule)                    62.4%  +0.0 pp  ✗ below target
  Lever 1: Redistribute (no new hires)           82.5%  +20.2 pp  ✓ TARGET MET
  Lever 2: Add 10% headcount                     76.9%  +14.5 pp  ✗ below target
  Lever 2b: Add 5 FTEs/interval (+35%)           95.5%  +33.1 pp  ✓ TARGET MET
  Lever 3: Reduce AHT 15%                        76.2%  +13.8 pp  ✗ below target
  Lever 4: Reduce shrinkage 4pts                 69.3%  +6.9 pp  ✗ below target
  Lever 5: Redistribute + AHT 15% + Shrinkage 4  94.4%  +32.0 pp  ✓ TARGET MET


In [21]:
# -- Module 6: Decision Lab visualization -------------------------------------

fig, axes = plt.subplots(1, 2, figsize=(14, 6))
fig.suptitle('Decision Lab — What-If Scenario Comparison', fontsize=14, fontweight='bold')

# Left: horizontal bar chart of SL by scenario
ax = axes[0]
labels = [s.replace('Lever ', 'L').split(':')[0] + ':' + s.split(':')[1] if ':' in s else s
          for s in scenarios['Scenario']]
labels = [s[:42] for s in scenarios['Scenario']]
sls    = scenarios['Service Level'] * 100
colors = ['#059669' if v >= TARGET_SL * 100 else '#1D4ED8' if i > 0 else '#DC2626'
          for i, v in enumerate(sls)]
bars = ax.barh(labels, sls, color=colors, edgecolor='white', linewidth=1)
ax.axvline(TARGET_SL * 100, color='#DC2626', ls='--', lw=1.5, label=f'80% target')
for bar, val in zip(bars, sls):
    ax.text(val + 0.3, bar.get_y() + bar.get_height()/2,
            f'{val:.1f}%', va='center', fontsize=9, fontweight='bold')
ax.set_xlim(0, 105)
ax.set_xlabel('Service Level (%)')
ax.set_title('Service Level by Scenario', fontweight='bold')
ax.legend(loc='lower right', fontsize=9)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.invert_yaxis()

# Right: delta from baseline
ax = axes[1]
deltas = scenarios['vs Baseline'].iloc[1:] * 100
dlabels = [s[:38] for s in scenarios['Scenario'].iloc[1:]]
dcolors = ['#059669' if v > 0 else '#DC2626' for v in deltas]
bars = ax.barh(dlabels, deltas, color=dcolors, edgecolor='white', linewidth=1)
ax.axvline(0, color='#6B7280', lw=1)
for bar, val in zip(bars, deltas):
    ax.text(val + 0.2, bar.get_y() + bar.get_height()/2,
            f'{val:+.1f} pp',
            va='center', fontsize=9, fontweight='bold')
ax.set_xlabel('Service level change vs baseline (percentage points)')
ax.set_title('SL Gain per Lever (vs Baseline)', fontweight='bold')
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.invert_yaxis()

plt.tight_layout()
plt.savefig('decision_lab_scenarios.png', dpi=150, bbox_inches='tight')
plt.show()
print('Chart saved: decision_lab_scenarios.png')


Chart saved: decision_lab_scenarios.png


**What the Decision Lab demonstrates:**

- Each lever is independently quantified — a WFM manager can compare the ROI of redistribution vs. hiring vs. AHT reduction before committing resources
- The combined lever scenario shows what is achievable by stacking operational improvements without new headcount
- Levers that don't move the needle enough to hit target are still valuable: they tell leadership *how far short* a given initiative falls, preventing over-investment in low-ROI fixes
- In production this runs as a live simulator — planners adjust sliders and see SL impact in real time

> *The Decision Lab reframes workforce planning from a reporting function into a decision-support function.*


---
## 9. Multichannel Staffing Methodology

Modern contact centers route work across three channel types — **voice, chat, and email** — each with fundamentally different queueing behavior. A WFM analyst who understands only voice is partially equipped. This section extends the methodology to all three channels, showing where the math changes and why it matters.

---

### Why channel type changes the staffing model

| | Voice | Chat | Email / Async |
|---|---|---|---|
| Queue model | Erlang C (1 agent per call) | Erlang C + concurrency | Backlog throughput model |
| Concurrency | 1 conversation per agent | 2–5 simultaneous chats | N/A (batch work) |
| SLA metric | % answered in X seconds | % answered in X seconds | % resolved within X hours |
| Staffing failure mode | Abandonment / long wait | Queue buildup at scale | Backlog accumulation |
| AHT definition | Talk + ACW per call | Handle time per chat session | Resolution time per ticket |

The key insight: **the same Erlang C engine powers voice and chat, but chat requires a concurrency adjustment before the offered load calculation.** Email bypasses queueing theory entirely — it is a throughput/backlog problem.


### 9.1 Chat Channel — Erlang C with Concurrency

**The mechanics:** A chat agent handles multiple simultaneous conversations. If the average concurrency is 3 (an agent manages 3 chats at once), the effective offered load *per agent slot* is lower than in voice — but the staffing math requires an explicit concurrency adjustment.

**Adjusted offered load:**

```
chat_erlangs_adjusted = chat_erlangs_raw / concurrency_factor
```

This adjusted figure is then passed into the standard Erlang C function. The result is agents-on-chat, which is then uplifted by the same shrinkage methodology as voice.

**Important caveat:** Higher concurrency lowers headcount but increases cognitive load per agent — quality degrades. Most contact centers cap concurrency at 3–4 for service-quality reasons, not technical ones.


In [22]:
# -- 9.1  Chat Channel Staffing Model ------------------------------------------
# SYNTHETIC ASSUMPTIONS -- labeled clearly
CHAT_VOLUME_PER_INTERVAL   = 45      # chat sessions arriving per 30-min interval
CHAT_AHT_SECONDS           = 480     # avg handle time per chat session
CHAT_CONCURRENCY           = 3.0     # avg simultaneous chats per agent (industry: 2-5)
CHAT_TARGET_SL             = 0.80    # 80% of chats answered within target
CHAT_TARGET_SEC            = 30      # within 30 seconds
CHAT_MAX_OCC               = 0.85    # lower ceiling for chat (cognitive load)
SHRINKAGE_RATE             = 0.34    # reuse voice shrinkage assumption

def agents_for_sl(A, target_sl, target_sec, aht_sec, max_occ):
    """Find minimum N agents satisfying SL and occupancy constraints."""
    import math
    N = max(1, math.ceil(A))
    while N < 500:
        occ = A / N
        if occ <= max_occ:
            pw = erlang_c_prob_wait(A, N)
            sl = 1 - pw * math.exp(-(N - A) * (target_sec / aht_sec))
            if sl >= target_sl:
                return N
        N += 1
    return N

# Step 1: Raw offered load
chat_raw_erlangs = (CHAT_VOLUME_PER_INTERVAL * CHAT_AHT_SECONDS) / (30 * 60)

# Step 2: Adjust for concurrency
chat_adj_erlangs = chat_raw_erlangs / CHAT_CONCURRENCY

# Step 3: Erlang C -- find minimum agents where SL target is met
chat_agents_on = agents_for_sl(chat_adj_erlangs, CHAT_TARGET_SL, CHAT_TARGET_SEC,
                               CHAT_AHT_SECONDS, CHAT_MAX_OCC)

# Step 4: Uplift to scheduled FTEs via shrinkage
import math
chat_ftes = math.ceil(chat_agents_on / (1 - SHRINKAGE_RATE))

# Compare to voice equivalent (same volume, no concurrency)
voice_equiv_erlangs = (CHAT_VOLUME_PER_INTERVAL * CHAT_AHT_SECONDS) / (30 * 60)
voice_equiv_agents  = agents_for_sl(voice_equiv_erlangs, CHAT_TARGET_SL, CHAT_TARGET_SEC,
                                    CHAT_AHT_SECONDS, 0.90)
voice_equiv_ftes    = math.ceil(voice_equiv_agents / (1 - SHRINKAGE_RATE))

print('=' * 55)
print('  CHAT CHANNEL STAFFING MODEL')
print('=' * 55)
print(f'  Volume per interval:     {CHAT_VOLUME_PER_INTERVAL} chat sessions')
print(f'  AHT per session:         {CHAT_AHT_SECONDS}s ({CHAT_AHT_SECONDS//60} min)')
print(f'  Raw offered load:        {chat_raw_erlangs:.2f} erlangs')
print(f'  Concurrency factor:      {CHAT_CONCURRENCY}x')
print(f'  Adjusted offered load:   {chat_adj_erlangs:.2f} erlangs')
print(f'  Agents on-chat required: {chat_agents_on}')
print(f'  Scheduled FTEs:          {chat_ftes}')
print()
print(f'  --- Concurrency impact ---')
print(f'  Same volume, no concurrency:  {voice_equiv_ftes} FTEs')
print(f'  With {CHAT_CONCURRENCY}x concurrency:       {chat_ftes} FTEs')
print(f'  Headcount reduction:     {voice_equiv_ftes - chat_ftes} FTEs ({(1 - chat_ftes/voice_equiv_ftes)*100:.0f}%)')
print()
print(f'  NOTE: {CHAT_CONCURRENCY}x concurrency assumes agents sustain quality')
print(f'     across simultaneous threads. Most centers cap at 3-4.')
print('=' * 55)


  CHAT CHANNEL STAFFING MODEL
  Volume per interval:     45 chat sessions
  AHT per session:         480s (8 min)
  Raw offered load:        12.00 erlangs
  Concurrency factor:      3.0x
  Adjusted offered load:   4.00 erlangs
  Agents on-chat required: 7
  Scheduled FTEs:          11

  --- Concurrency impact ---
  Same volume, no concurrency:  25 FTEs
  With 3.0x concurrency:       11 FTEs
  Headcount reduction:     14 FTEs (56%)

  NOTE: 3.0x concurrency assumes agents sustain quality
     across simultaneous threads. Most centers cap at 3-4.


### 9.2 Email / Async Channel — Backlog Throughput Model

Email does not queue in real time — it accumulates. There is no 'abandonment' and Erlang C does not apply. The staffing question is: **how many agents do I need to work through today's volume and any carryover backlog within the SLA window?**

**Model structure:**

```
total_workload     = carryover_backlog + new_volume_today
throughput/agent   = productive_hours * 60 / handle_time_per_ticket
agents_required    = ceil(total_workload / throughput_per_agent)
```

**The staffing failure mode is different from voice:** You do not see it in real-time queue depth. You see it at end-of-day when backlog has grown and SLA compliance has slipped. This is why email channels are chronically understaffed — the pain is deferred and invisible until the backlog compounds.


In [23]:
# -- 9.2  Email / Async Backlog Throughput Model --------------------------------
# SYNTHETIC ASSUMPTIONS -- labeled clearly
EMAIL_DAILY_VOLUME         = 320    # new email tickets arriving today
EMAIL_CARRYOVER_BACKLOG    = 85     # unresolved tickets from prior day
EMAIL_SLA_HOURS            = 24     # target: resolve within 24 hours
EMAIL_HANDLE_TIME_MIN      = 8      # avg resolution time per ticket (minutes)
PRODUCTIVE_HOURS_PER_AGENT = 6.0    # usable hours per 8-hr shift after shrinkage

# Step 1: Total workload
total_tickets = EMAIL_DAILY_VOLUME + EMAIL_CARRYOVER_BACKLOG

# Step 2: Throughput per agent per day
tickets_per_agent_per_day = (PRODUCTIVE_HOURS_PER_AGENT * 60) / EMAIL_HANDLE_TIME_MIN

# Step 3: Agents required
email_agents_required = math.ceil(total_tickets / tickets_per_agent_per_day)

# Step 4: Scheduled FTEs (shrinkage already baked into PRODUCTIVE_HOURS_PER_AGENT)
email_ftes = email_agents_required

# Backlog compounding scenario -- what if we understaff by 20%?
understaffed_agents = math.floor(email_agents_required * 0.80)
tickets_resolved    = math.floor(understaffed_agents * tickets_per_agent_per_day)
end_of_day_backlog  = total_tickets - tickets_resolved
next_day_total      = EMAIL_DAILY_VOLUME + end_of_day_backlog
next_day_agents     = math.ceil(next_day_total / tickets_per_agent_per_day)

print('=' * 55)
print('  EMAIL / ASYNC CHANNEL STAFFING MODEL')
print('=' * 55)
print(f'  New volume today:        {EMAIL_DAILY_VOLUME} tickets')
print(f'  Carryover backlog:       {EMAIL_CARRYOVER_BACKLOG} tickets')
print(f'  Total workload:          {total_tickets} tickets')
print(f'  Handle time per ticket:  {EMAIL_HANDLE_TIME_MIN} min')
print(f'  Productive hrs/agent:    {PRODUCTIVE_HOURS_PER_AGENT} hrs')
print(f'  Throughput/agent/day:    {tickets_per_agent_per_day:.0f} tickets')
print(f'  Agents required:         {email_agents_required}')
print(f'  Scheduled FTEs:          {email_ftes}')
print()
print('  --- Backlog compounding risk (80% staffing scenario) ---')
print(f'  Agents deployed:         {understaffed_agents}')
print(f'  Tickets resolved today:  {tickets_resolved}')
print(f'  End-of-day backlog:      {end_of_day_backlog} tickets')
print(f'  Tomorrow total load:     {next_day_total} tickets')
print(f'  Tomorrow agents needed:  {next_day_agents} (+{next_day_agents - email_agents_required} vs today)')
print()
print('  NOTE: Email understaffing compounds silently.')
print('     Pain shows in SLA reports, not queue dashboards.')
print('=' * 55)


  EMAIL / ASYNC CHANNEL STAFFING MODEL
  New volume today:        320 tickets
  Carryover backlog:       85 tickets
  Total workload:          405 tickets
  Handle time per ticket:  8 min
  Productive hrs/agent:    6.0 hrs
  Throughput/agent/day:    45 tickets
  Agents required:         9
  Scheduled FTEs:          9

  --- Backlog compounding risk (80% staffing scenario) ---
  Agents deployed:         7
  Tickets resolved today:  315
  End-of-day backlog:      90 tickets
  Tomorrow total load:     410 tickets
  Tomorrow agents needed:  10 (+1 vs today)

  NOTE: Email understaffing compounds silently.
     Pain shows in SLA reports, not queue dashboards.


### 9.3 Three-Channel Summary Comparison

Bringing all three channels together shows how a WFM analyst thinks about blended workforce planning: each channel has different volume patterns, different SLA mechanics, and different failure modes — but the same shrinkage and scheduling constraints apply across all of them.


In [24]:
# -- 9.3  Three-Channel Summary Chart ------------------------------------------
import pandas as pd
import matplotlib.pyplot as plt

# Representative voice peak-interval figures
VOICE_PEAK_ERLANGS = 8.5
VOICE_PEAK_AGENTS  = agents_for_sl(VOICE_PEAK_ERLANGS, 0.80, 20, 240, 0.90)
VOICE_PEAK_FTES    = math.ceil(VOICE_PEAK_AGENTS / (1 - SHRINKAGE_RATE))

fig, axes = plt.subplots(1, 2, figsize=(14, 5))
fig.suptitle('Multichannel Staffing: Model Comparison', fontsize=14, fontweight='bold')

# Left: FTE bar chart
channels = ['Voice\n(peak interval)', 'Chat\n(3x concurrency)', 'Email/Async\n(daily)']
ftes     = [VOICE_PEAK_FTES, chat_ftes, email_ftes]
colors   = ['#1D4ED8', '#0891B2', '#7C3AED']
bars = axes[0].bar(channels, ftes, color=colors, width=0.5, edgecolor='white', linewidth=1.5)
for bar, val in zip(bars, ftes):
    axes[0].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.2,
                 str(val), ha='center', va='bottom', fontweight='bold', fontsize=13)
axes[0].set_title('Scheduled FTEs by Channel', fontweight='bold')
axes[0].set_ylabel('Scheduled FTEs')
axes[0].set_ylim(0, max(ftes) * 1.35)
axes[0].spines['top'].set_visible(False)
axes[0].spines['right'].set_visible(False)

# Right: Risk/visibility matrix as table
axes[1].axis('off')
table_data = [
    ['Channel',     'Queue Model',              'SLA Metric',         'Failure Visible In'],
    ['Voice',       'Erlang C',                 '% ans / 20s',        'Real-time queue'],
    ['Chat',        'Erlang C + Concurrency',   '% ans / 30s',        'Real-time queue'],
    ['Email/Async', 'Backlog Throughput',        '% resolved / 24h',  'EOD SLA report'],
]
col_w  = [0.18, 0.30, 0.26, 0.26]
row_bg = ['#1E3A5F', '#EFF6FF', '#E0F2FE', '#F5F3FF']
row_fg = ['white',   '#1E293B', '#1E293B', '#1E293B']
y = 0.82
for r, (row, bg, fg) in enumerate(zip(table_data, row_bg, row_fg)):
    x = 0.01
    for cell, cw in zip(row, col_w):
        axes[1].add_patch(plt.Rectangle((x, y-0.13), cw-0.01, 0.14,
                          transform=axes[1].transAxes, color=bg, zorder=1))
        axes[1].text(x + (cw-0.01)/2, y-0.055, cell,
                    transform=axes[1].transAxes, ha='center', va='center',
                    fontsize=8.5, fontweight='bold' if r==0 else 'normal',
                    color=fg, zorder=2)
        x += cw
    y -= 0.17
axes[1].set_title('Channel Risk & Visibility Matrix', fontweight='bold')
axes[1].set_xlim(0, 1)
axes[1].set_ylim(0, 1)

plt.tight_layout()
plt.savefig('multichannel_comparison.png', dpi=150, bbox_inches='tight')
plt.show()
print('Chart saved: multichannel_comparison.png')


Chart saved: multichannel_comparison.png


### Key Takeaways

1. **Voice and chat share the same queueing engine (Erlang C)** — but chat requires a concurrency adjustment before the offered load calculation. Skipping this step overstates headcount and makes chat look more expensive than it is.

2. **Email is a fundamentally different problem.** There is no queue — there is a backlog. Erlang C does not apply. Understaffing an email team is invisible in real-time dashboards and only surfaces in end-of-day SLA reports — by which point the backlog has already compounded.

3. **The same shrinkage methodology applies across all channels.** Breaks, training, coaching, and PTO exist regardless of channel. The uplift from agents-on-channel to scheduled FTEs is the same calculation.

4. **Production WFM tools (NICE, Verint, Calabrio, Genesys) automate this multichannel math.** This section demonstrates the underlying methodology — the reasoning the tools encode.


## 8. Summary

This notebook runs the full WFM planning chain on real interval-level arrival data:

1. **Forecast** — interval-level call volume via a transparent profile-ratio method, validated on a holdout and reported with the volume-weighted error metric (WAPE) that WFM teams actually track.
2. **Workload** — converted calls to offered load using an explicit AHT assumption.
3. **Erlang C** — sized agents-on-phone to an 80/20 service-level target under a 90% occupancy ceiling.
4. **Shrinkage** — uplifted on-phone requirement to scheduled FTEs with a fully decomposed 34% shrinkage build-up.
5. **Gap analysis** — compared required vs. a synthetic lumpy shift schedule, surfaced over/under-staffed intervals, and showed delivered service level interval-by-interval.
6. **Redistribution (Module 5)** — moved surplus agent-intervals into deficit intervals, holding the total constant: 62.4% → 82.5% (+20.2 pp).
7. **Scenario comparison (Module 6)** — redistribution vs a 10% headcount add, AHT and shrinkage levers, and a combined scenario.

**Methodology, not platform.** The same logic powers commercial WFM suites (NICE, Verint, Calabrio). This notebook demonstrates the underlying methodology directly in Python — the forecasting, queueing, and capacity-planning reasoning that sits beneath any tool.

**Extensions a production build would add:** Abandonment-aware queueing (Erlang A/X), intraday re-forecasting, multi-skill / blended workload, shift-bidding optimization, and a what-if sensitivity layer for AHT and shrinkage. Section 9 demonstrates the full voice → chat → email multichannel planning chain.

---

*Data: "Anonymous Bank" call-center dataset, SEE Lab, Technion (Prof. Avi Mandelbaum), via the Rfssa data mirror. WFM overlay (AHT, shrinkage, service-level target, schedule) is synthetic and labeled throughout. Business logic and operational framing by the author; AI assisted with Python syntax and plotting.*
